### PR TITLE
Allocate minimum size necessary in pango text functions.

### PR DESCRIPTION
**Pull request by c-edw**
Almost forgot about this. See [here](https://github.com/swaywm/sway/pull/3438/files/338a8c051df8290eb7863ba0c566139747b7b224#r248228815) (annoyingly this doesn't link directly to the comment. It's in one of the resolved conversations...).

### CODE REVIEW

File: common/pango.c [hunk 1]
```diff
@@ -88,14 +89,22 @@ PangoLayout *get_pango_layout(cairo_t *cairo, const char *font,
...
 	va_list args;
 	va_start(args, fmt);
-	if (vsnprintf(buf, sizeof(buf), fmt, args) >= MAX_CHARS) {
-		strcpy(&buf[sizeof(buf) - sizeof(overflow)], overflow);
-	}
+	// Add one since vsnprintf excludes null terminator.
+	int length = vsnprintf(NULL, 0, fmt, args) + 1;
 	va_end(args);
+	// Limit length of string to MAX_CHARS.
+	int length_limited = imin(length, MAX_CHARS);
```
Comment (emersion) [inline common/pango.c]: TBH I don't think limiting to `MAX_CHARS` is necessary.

Comment (c-edw) [reply common/pango.c]: Unlimited size? I guess the only benefit is protection against oversized strings created from a bug or something like that.

Comment (emersion) [reply common/pango.c]: Yeah, we don't do this elsewhere, let's not do it here.

File: .gitignore [hunk 2]
```diff
@@ -5,6 +5,8 @@ install_manifest.txt
 bin/
 test/
 build/
+build-gcc/
```
Comment (c-edw) [inline .gitignore]: May I add these? It makes things easier when testing for each compiler.

**Comment by c-edw** [comment]
Anomaly is causing CI build failure. Code compiles.

```
...
(1/5) Probing GDK-Pixbuf loader modules...
(2/5) Reloading system manager configuration...
(3/5) Reloading device manager configuration...
(4/5) Arming ConditionNeedsUpdate...
(5/5) Updating the MIME type database...
:: Searching databases for updates...
:: Searching AUR for updates...
Get https://aur.archlinux.org/rpc.php?arg%5B%5D=yay-bin&type=info&v=5: dial tcp 5.9.250.164:443: i/o timeout
```

File: include/util.h [hunk 3]
```diff
@@ -8,7 +8,12 @@
 #include <xkbcommon/xkbcommon.h>
 
 /**
- * Wrap i into the range [0, max[
+ * Return the minimum (lowest value) of two integers.
+ */
+int imin(int a, int b);
```
Comment (mstoeckl) [inline include/util.h]: This function is not used anywhere?

File: .gitignore (hunk 2 above)
Comment (emersion) [reply .gitignore]: Not a fan of this.

Comment (c-edw) [reply .gitignore]: What can I do instead? Without this, I was having to do:

```
CC=gcc meson build
rm -rf build
...
CC=clang meson build
...
rm -rf build
```

Comment (emersion) [reply .gitignore]: I'd prefer something like `/build-*` instead.

**Comment by c-edw** [comment]
Yep, just force pushed.

File: include/util.h [hunk 4]
```diff
@@ -8,7 +8,7 @@
 #include <xkbcommon/xkbcommon.h>
 
 /**
- * Wrap i into the range [0, max[
+ * Wrap i into the range [0, max]
```
Comment (emersion) [inline include/util.h]: I think "[0, max[" is correct.

Unrelated to this PR: I wonder why it's a function actually? Isn't `((i % max) + max) % max` the same as `i % max`?

Comment (c-edw) [reply include/util.h]: I believe that in C, `%` calculates the remainder. `((i % max) + max) % max` calculates the modulus.

Comment (emersion) [reply include/util.h]: `%` _is_ the modulus operator…

File: common/pango.c [hunk 5]
```diff
@@ -88,13 +84,15 @@ PangoLayout *get_pango_layout(cairo_t *cairo, const char *font,
...
 	va_list args;
 	va_start(args, fmt);
-	if (vsnprintf(buf, sizeof(buf), fmt, args) >= MAX_CHARS) {
-		strcpy(&buf[sizeof(buf) - sizeof(overflow)], overflow);
-	}
+	// Add one since vsnprintf excludes null terminator.
+	int length = vsnprintf(NULL, 0, fmt, args) + 1;
+	va_end(args);
+
+	char *buf = malloc(length);
```
Comment (ddevault) [inline common/pango.c]: Need to handle malloc failure

File: common/pango.c [hunk 6]
```diff
@@ -104,17 +102,20 @@ void get_text_size(cairo_t *cairo, const char *font, int *width, int *height,
...
```
Comment (ddevault) [inline common/pango.c]: Here as well

File: common/pango.c [hunk 7]
```diff
@@ -88,13 +84,15 @@ PangoLayout *get_pango_layout(cairo_t *cairo, const char *font,
...
 		int *baseline, double scale, bool markup, const char *fmt, ...) {
-	char buf[MAX_CHARS];
-
 	va_list args;
 	va_start(args, fmt);
-	if (vsnprintf(buf, sizeof(buf), fmt, args) >= MAX_CHARS) {
-		strcpy(&buf[sizeof(buf) - sizeof(overflow)], overflow);
-	}
+	// Add one since vsnprintf excludes null terminator.
+	int length = vsnprintf(NULL, 0, fmt, args) + 1;
```
Comment (emersion) [inline common/pango.c]: Nitpick: I prefer to use "length" (or "len") to mean the return value of `strlen` (aka without the terminating NULL byte) and "size" to mean the value passed to `malloc` et al (aka including the terminating NULL byte).

**Comment by c-edw** [comment]
In what cases do allocation failures need to be handled? Is it always? If that's the case, is that something that wasn't enforced in the past but is now?

File: common/pango.c (hunk 7 above)
Comment (c-edw) [reply common/pango.c]: Yep, that's fair. I can change it if you really want.

**Comment by ddevault** [comment]
Just where it's important. In this case it's possibly exploitable by an evil client which sets its title to something ridiculous that we can't draw.

**Comment by emersion** [comment]
In a perfect world, the compositor would survive even if the device is running low on memory.

**Comment by emersion** [comment]
Thanks!

**Pull request merged**